packages/@ember/-internals/runtime/lib/system/array_proxy.ts

```diff
@@ -183,30 +218,33 @@ export default class ArrayProxy extends EmberObject {
     @method replaceContent
     @param {Number} idx The starting index
     @param {Number} amt The number of items to remove from the content.
-    @param {EmberArray} objects Optional array of objects to insert or null if no
-      objects.
+    @param {EmberArray} objects Optional array of objects to insert.
     @return {void}
     @public
   */
-  replaceContent(idx, amt, objects) {
-    get(this, 'content').replace(idx, amt, objects);
+  replaceContent(idx: number, amt: number, objects?: T[]) {
+    let content = get(this, 'content');
+    assert('[BUG] Called objectAtContent without content', content);
+    assert('Mutating a non-mutable array is not allowed', isMutable(content));
+    content.replace(idx, amt, objects);
   }
 
   // Overriding objectAt is not supported.
-  objectAt(idx) {
+  objectAt(idx: number) {
     this._revalidate();
 
     if (this._objects === null) {
       this._objects = [];
     }
 
     if (this._objectsDirtyIndex !== -1 && idx >= this._objectsDirtyIndex) {
-      let arrangedContent = get(this, 'arrangedContent');
+      let arrangedContent = get(this, 'arrangedContent') as C | null;
       if (arrangedContent) {
+        // SAFETY: The type of length should be inferred as number but is not for unknown reasons.
         let length = (this._objects.length = get(arrangedContent, 'length'));
 
         for (let i = this._objectsDirtyIndex; i < length; i++) {
-          this._objects[i] = this.objectAtContent(i);
+          this._objects[i] = this.objectAtContent(i)!;
```

Comment:
A `// SAFETY` comment here would be great. I assume it's because there's an invariant relating the loop to `objectAtContent`?